Added Tests for telemetry test tool (including bug fixes)

JannisW opened the pull request:
# General changes

1. Building process changed as two executables have to be created (google test suite and actual test tool, resulting in major structural changes)
2. Test cases added in TelemetryReportingTest.cpp (required creation of interface and mock classes)
3. Bug fixes of found problems by tests
4. Specification updated due to errors (https://docs.google.com/spreadsheets/d/1soVLjeD9Sl7z7Z6cYMyn1fmn-cG7tx_pfFDsvgkCqMU/edit#gid=782574835)

# Bugs
## General
1. Wrong / imprecise specification: Wrong package sizes, because of mising meta data byte _(FIXED)_
2. stuffData() fails if a package of maximal COBS length (254 bytes) and without any zero byte in payload, should be sent. The result is a non zero termination byte. 
_(OPEN: suggested fix delete line 465 in TelemetryReporting.cpp which increases lengthOfEncodedData if max length is reached; just if this is not intended )_ 
(However this bug is not severe as all messages are shorter than 254 bytes)

## sendDriverControls
Wrong package length in constant (value: 10) _(FIXED: now it has the correct value of 9)_

## sendBatteryData
Data for BatteryData at position 43 and 47 (current, power) was not written _(FIXED: new fileds in BatteryData.h, new Method writeUIntIntoArray in TelemetryReporting.cpp and new field in DataUnion also TelemetryReporting.cpp)_

## sendCmu
CMU "Cell X Voltage" type is short uint but Note in specification says "Negative indicates channel mismatch" _(FIXED: offset 2 to 16 changed from short uint to signed int in specifacation. Also changed the code: changed type of cellVoltage to signed short in CmuData.h, changed method to writeShortIntoArray in TelemetryReporting.cpp)_

## sendMppt
1. Alive bit was not considered _(FIXED: changed code in TelemetryReporting.cpp)_
2. Wrong package length was used (KEY_MOTOR_LENGTH) _(FIXED: now uses MPPT_LENGTH)_
3. Wrong package id was used (KEY_MOTOR_PKG_ID) _(FIXED: now uses MPPT_PKG_ID)_

## sendLights
1. Wrong package length was used (KEY_MOTOR_LENGTH) _(FIXED: now uses LIGHTS_LENGTH)_
2. Wrong package id was used (KEY_MOTOR_PKG_ID) _(FIXED: now uses LIGHTS_PKG_ID)_

## sendMotorDetails
The signal / slot configuration was not correct. The buttons weren't correctly connected with the actual parameterized method in TelemetryReporting.cpp

enochtsang commented on src/Tests/BusinessLayer/TelemetryReportingTest.cpp:
```diff
@@ -0,0 +1,959 @@
+
+    expectedMax[0] = 0xFF;
+    expectedMax[MAX_COBS_PACKAGE_LENGTH + 1] = 0x00;
+    assertThatCOBScorrect(inputMax, expectedMax, MAX_COBS_PACKAGE_LENGTH); // TODO this one fails. can probably be fixed by removing line 452 from TelemetryReporting.cpp
```
You know I think you were right about removing that line. If we have the whole testing infrastructure to make sure stuff still works, just remove that line and fix this error.

JannisW replied:
sounds good! til then I'll keep the line

enochtsang commented on src/Tests/BusinessLayer/TelemetryReportingTest.cpp:
```diff
@@ -0,0 +1,959 @@
+
+/*
+ * This test tests if the checksum is correctly appended. It does not test the checksum calculation itself.
+ * (Supposed to work correctly, as copied from already tested code?!)
```
Yup, this is true, this file doesn't need to test the CRC Calculator. You can remove this line.

JannisW commented on src/Tests/BusinessLayer/TelemetryReportingTest.cpp:
```diff
@@ -0,0 +1,959 @@
+                                                        ));
+    }
+
+    void assertThatCOBScorrect(unsigned char* input, unsigned char* expected, const unsigned int inputLength)
```
I changed it to expectCobsCorrect as I changed the use of ASSERT_THAT within the method to EXPECT_THAT

JannisW commented on src/Tests/BusinessLayer/TelemetryReportingTest.cpp:
```diff
@@ -0,0 +1,959 @@
+    telemetryReporting_->writeFloatIntoArray(data, 39, keyMotorData_->motor1VehicleVelocity);
+    appendChecksum(data, payloadLength);
+    // do some additional data checks
+    ASSERT_THAT(data[0], Eq(0x01)); // packet id
```
Yes, this is redundant now. I inserted these kind of assertions in the first place to ensure that I didn't introduce bugs during writing my test cases. With respect to that it makes sense that these are asserts rather than expects as the test case itself would be wrong.
It also checks (kind of) if the appendChecksum method affects the array at positions other than the last ones.

JannisW commented on src/Tests/BusinessLayer/TelemetryReportingTest.cpp:
```diff
@@ -0,0 +1,959 @@
+    // do some additional data checks
+    ASSERT_THAT(data[0], Eq(0x01)); // packet id
+    ASSERT_THAT(data[1], Eq(0x01));
+    ASSERT_THAT(data[22], Eq(0x00));
```
made it explicit by using constants

JannisW commented on src/Tests/BusinessLayer/TelemetryReportingTest.cpp:
```diff
@@ -0,0 +1,959 @@
+    unsigned char expectedPacket[expectedPackageLength];
+    telemetryReporting_->frameData(data, payloadLength, expectedPacket);
+    // check call
+    EXPECT_CALL(*communicationService_, sendData(_, expectedPackageLength)).With(Args<0, 1>(ElementsAreArray(expectedPacket, expectedPackageLength))).Times(1);
```
although it is a pretty long line I still think it is quite readable because of the powerful google test matcher mechanism.
However I extracted a part into a local variable splitting the one liner into two shorter parts

JannisW commented on src/BusinessLayer/TelemetryReporting.h:
```diff
@@ -33,29 +35,58 @@ class TelemetryReporting : public QObject
                        const LightsData& lightsData_,
                        View& view);
 
+    friend class TelemetryReportingTest;
```
I don't really get how this is suppose to help us here. To test the correct behavior it is required to call the private function in order to check if the mocked communication service is called correctly. I don't see how we could avoid this by using mocked data classes.
However creation of mocked data classes make sense as soon as they provide additional methods. So far they just store public data. Therefore one could simply initialize the used fields explicitely in the setUp method. Do you think I should do that?

enochtsang replied:
We shouldn't need to call private functions, even in our tests. By adding code in here specifically for testing, we are also hurting the portability of this code. If we ever wanted to move it somewhere, google test would have to be added to that project. 

Testing should be done by using the entirety of the public interface, doing this should test all the private functionality as well. When the mocked Data classes are passed into the constructor of this class, this will test the private slots because TelemetryReporting does the job of connecting them.

For the private functions like framing data and stuff, this was a design flaw from before you touched this project (i.e. my fault ;D). All the private functions are too much responsibility for this one class and should be moved to a different class where the functions can be public and properly tested.

JannisW commented:
Right now the travis build fails just because one test case is failing (the one mentioned earlier, which can probably be fixed by removing the marked line).

enochtsang commented:
Yea just go ahead and remove it Jannis, let's make sure the build is passing.

JannisW commented:
Found and fixed a new bug (sendMotorDetails: The signal / slot configuration was not correct. The buttons weren't correctly connected with the actual parameterized method in TelemetryReporting.cpp)

Fixed the friend class problem by calling the functions using qt's signal and slot mechanism
